### PR TITLE
driver metadata changes

### PR DESCRIPTION
**Pull request by kinarashah**
- sync only if interval!=0 and hash changes
- git clone instead of using raw github content
- updated log msgs

https://github.com/rancher/rancher/issues/22483
https://github.com/rancher/rancher/issues/22329

### CODE REVIEW

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 1]
```diff
@@ -284,3 +268,115 @@ func loadData(url string) (Data, error) {
...
+	if err != nil {
+		return data, err
+	}
+
+	if err := json.Unmarshal(buf, &data); err != nil {
+		return data, fmt.Errorf("error unmarshaling metadata contents %v", err)
+	}
+
+	if err := os.RemoveAll(path); err != nil {
+		logrus.Errorf("err removing %v", err)
```
Comment (mrajashree) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: nit: can we add path name to error message?

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 2]
```diff
@@ -74,52 +98,56 @@ func Register(ctx context.Context, management *config.ManagementContext) {
...
-			prevURL = currURL
-		}
-		return setting, nil
+	settingValues, err := getSettingValues()
+	if err != nil {
+		return nil, err
 	}
 
-	interval, err := getTimeSetting()
+	interval, err := parseTime(settingValues[refreshInterval])
```
Comment (mrajashree) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: I think it makes sense to do parseTime check in a store create/update in addition to this as well. So even if the time is negative/zero controller won't keep returning error

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 3]
```diff
@@ -6,9 +6,14 @@ import (
...
 	"io/ioutil"
 	"net/http"
+	"os"
+	"path/filepath"
 	"strings"
+	"sync"
 	"time"
 
+	"github.com/rancher/rancher/pkg/randomtoken"
+
```
Comment (mrajashree) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: nit: can we remove extra blank line :)

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 4]
```diff
@@ -74,52 +98,56 @@ func Register(ctx context.Context, management *config.ManagementContext) {
...
-			logrus.Debugf("driverMetadata: canceled counter")
-			tickerData.cancelFunc()
-			tickerData.interval = 0
+	url, urlErr := parseURL(settingValues)
+
+	if tickerData == nil {
+		// load from stored metadata on error
+		if interval == 0 || urlErr != nil {
+			logrus.Errorf("driverMetadata: error loading data, using stored defaults: interval %v, error %v", interval, urlErr)
+			if err := m.createOrUpdateMetadataDefaults(); err != nil {
```
[the remark this answers is not in the document]
Comment (kinarashah) [reply pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: I think the general idea is we won't make use of URL if the refresh is disabled(by setting interval 0), so even if url is changed, we'd consider it as a disabled action

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 5]
```diff
@@ -74,52 +98,56 @@ func Register(ctx context.Context, management *config.ManagementContext) {
...
+	interval, err := parseTime(settingValues[refreshInterval])
 	if err != nil {
 		if err.Error() != "refresh disabled" {
 			return nil, err
 		}
 	}
 
-	if tickerData.interval != interval {
+	// don't sync if time is set to negative/zero
+	if interval == 0 && tickerData != nil && tickerData.interval != 0 {
```
[the remark this answers is not in the document]
Comment (kinarashah) [reply pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: same as below, 
> I think the general idea is we won't make use of URL if the refresh is disabled(by setting interval 0), so even if url is changed, we'd consider it as a disabled action

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go (hunk 2 above)
Comment (kinarashah) [reply pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: we don't keep returning error if time is negative/zero. but adding validation in store seems like a good idea, would add it in a separate pr?

File: pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go [hunk 6]
```diff
@@ -130,102 +153,77 @@ func (m *MetadataController) sync(key string, setting *v3.Setting) (runtime.Obje
...
-			logrus.Errorf("error loading rke data, using stored defaults %v", err)
-			if err := m.createOrUpdateMetadataDefaults(); err != nil {
-				return err
-			}
-			return nil
-		}
-		return err
+func (m *MetadataController) refresh(url *URL, init bool) error {
+	if !toSync(url) {
+		logrus.Infof("driverMetadata: skip sync, hash up to date %v", url.latestHash)
```
Comment (alena1108) [inline pkg/controllers/management/kontainerdrivermetadata/metadata_handler.go]: shall we log is in debug? or having it in info is useful

**Pull request merged**